### PR TITLE
tests: assert Grace IO virtualization rejection on unsupported architectures

### PR DESCRIPTION
**Pull request by vamsikrishna-siddu**
#### Before this PR:
The `should set Grace conversion flags and expected aliases from options` unit test in `pkg/virt-launcher/virtwrap` fails on s390x because it exercises GraceIOVirtualization, which requires PCIe placement a capability s390x does not support. The test only asserted the success path, so it broke on architectures where `SupportPCIePlacement()` returns false.

#### After this PR:
The test branches on `arch.NewConverter(runtime.GOARCH).SupportPCIePlacement()` and asserts the expected behavior for both supported and unsupported architectures, fixing the `pull-kubevirt-unit-test-s390x` CI lane.

### Why we need it and why it was done in this way
The production code in `converter/grace.go` already rejects GraceIOVirtualization on architectures where `SupportPCIePlacement()` returns false. The test was only asserting the success path, which made it fail on s390x. By branching on the same capability check, the test now validates the success path on supported architectures (amd64, arm64) and the expected rejection on unsupported ones (s390x).

The following tradeoffs were made:

The following alternatives were considered:

Links to places where the discussion took place: 

### Special notes for your reviewer
`generateConverterContext` internally uses `arch.NewConverter(runtime.GOARCH)` to select the architecture converter, so the test is implicitly coupled to the host architecture. A broader refactor to inject the architecture into `generateConverterContext` would decouple this, but that is out of scope for this fix.

### Checklist

This checklist is not enforcing, but it's a reminder of items that could be relevant to every PR.
Approvers are expected to review this list.

- [ ] Design: A [VEP (virtualization enhancement proposal)](https://github.com/kubevirt/enhancements/blob/main/README.md) was considered and is present (link) or not required
- [ ] PR: The PR description is expressive enough and will help future contributors
- [ ] Code: [Write code that humans can understand](https://en.wikiquote.org/wiki/Martin_Fowler#code-for-humans) and [Keep it simple](https://en.wikipedia.org/wiki/KISS_principle)
- [ ] Refactor: You have [left the code cleaner than you found it (Boy Scout Rule)](https://learning.oreilly.com/library/view/97-things-every/9780596809515/ch08.html)
- [ ] Upgrade: Impact of this change on upgrade flows was considered and addressed if required
- [ ] Testing: New code requires [new unit tests](https://github.com/kubevirt/kubevirt/blob/main/docs/reviewer-guide.md#when-is-a-pr-good-enough). New features and bug fixes require at least one e2e test
- [ ] Documentation: A [user-guide update](https://github.com/kubevirt/user-guide/) was considered and is present (link) or not required. You want a user-guide update if it's a user facing feature / API change.
- [ ] Community: Announcement to [kubevirt-dev](https://groups.google.com/g/kubevirt-dev/) was considered
- [ ] AI Contributions: The PR abides by the [KubeVirt AI Contribution Policy](https://github.com/kubevirt/community/blob/main/ai-contribution-policy.md).

### Release note

```release-note
     None.

```

### CODE REVIEW

**Comment by suPer8Hu** [comment]
The skip looks correct to me, GraceIOVirtualization is rejected when SupportPCIePlacement() is false and s390x intentionally returns false there. Matching the existing GOARCH skip pattern in this file is fine. Can try to retest it~

**Review by dollierp — APPROVED** [review]
/lgtm

**Comment by enp0s3** [comment]
Hi @fanzhangio 
It looks like #18474 introduced regression on s390x. Can you please check the tests in `manager_test.go` that need proper arch injection?

**Comment by enp0s3** [comment]
Here is a reference to a failure: https://prow.ci.kubevirt.io/view/gs/kubevirt-prow/pr-logs/pull/kubevirt_kubevirt/18460/pull-kubevirt-unit-test-s390x/2081975713707069440

**Comment by enp0s3** [comment]
/approve

This deserves refactoring, but it will be out of this PR scope. Will take it in to solve the s390x regression.

**Comment by fossedihelm** [comment]
How about do something like:
```golang
		It("should set Grace conversion flags and expected aliases from options on supported architectures", func() {
			vmi.Spec.Domain.CPU = &v1.CPU{DedicatedCPUPlacement: true}
			iommuFDFile, err := os.CreateTemp(GinkgoT().TempDir(), "iommufd-test")
			Expect(err).ToNot(HaveOccurred())
			DeferCleanup(iommuFDFile.Close)

			options := &cmdv1.VirtualMachineOptions{
				VirtualMachineSMBios: &cmdv1.SMBios{},
				ClusterConfig: &cmdv1.ClusterConfig{
					GraceIOVirtualizationEnabled: true,
					PCINUMAAwareTopologyEnabled:  true,
				},
				GraceHostDeviceAliases: []string{"gpu-gpu0"},
			}

			libvirtManager := manager.(*LibvirtDomainManager)
			libvirtManager.iommuFD = int(iommuFDFile.Fd())
			converterContext, err := libvirtManager.generateConverterContext(vmi, true, options, false)

			if arch.NewConverter(runtime.GOARCH).SupportPCIePlacement() {
				Expect(err).ToNot(HaveOccurred())
				Expect(converterContext.GraceIOVirtualizationEnabled).To(BeTrue())
				Expect(converterContext.GraceHostDeviceAliases).To(Equal([]string{"gpu-gpu0"}))
			} else {
				Expect(err).To(MatchError(ContainSubstring("requires PCIe placement support")))
			}
		})
```
@enp0s3 WDYT?

**Comment by enp0s3** [comment]
@fossedihelm Thank you! LGTM

**Comment by enp0s3** [comment]
/hold
@vamsikrishna-siddu My apologies for the interrupt, would you be able to update the PR according to @fossedihelm suggestion?

File: pkg/virt-launcher/virtwrap/manager_test.go [hunk 1]
```diff
@@ -3259,6 +3259,9 @@ var _ = Describe("Manager", func() {
 		})
 
 		It("should set Grace conversion flags and expected aliases from options", func() {
+			if runtime.GOARCH == "s390x" {
+				Skip("GraceIOVirtualization requires PCIe placement which s390x does not support") //nolint:forbidigo
+			}
```
Comment (akalenyu) [inline pkg/virt-launcher/virtwrap/manager_test.go]: Not sure I get why a unit test for this feature fails?
Does the unit test somehow exercise the "real" architecture/logic of the host running it?

I feel like there's a deeper issue in here in that we don't control the arch path and instead rely on the host running the tests. Somewhat reminds me of this old discussion: https://github.com/kubevirt/kubevirt/pull/15199#discussion_r2212667825

**Comment by vamsikrishna-siddu** [comment]
> /hold @vamsikrishna-siddu My apologies for the interrupt, would you be able to update the PR according to @fossedihelm suggestion?

Hi @enp0s3 sure no problem. I will update the PR as suggest by @fossedihelm

File: pkg/virt-launcher/virtwrap/manager_test.go (hunk 1 above)
Comment (Barakmor1) [inline pkg/virt-launcher/virtwrap/manager_test.go]: Hey wouldn't it be better to fail instead?

File: pkg/virt-launcher/virtwrap/manager_test.go [hunk 2]
```diff
@@ -3259,6 +3259,9 @@ var _ = Describe("Manager", func() {
 		})
 
 		It("should set Grace conversion flags and expected aliases from options", func() {
```
Comment (Barakmor1) [inline pkg/virt-launcher/virtwrap/manager_test.go]: Is there a decorator that suggest that tests don't work with s390x?

**Review by akalenyu — COMMENTED** [review]
Thank you for tackling this! I have some worries about the testing approach
(which isn't necessarily going to converge in this PR) but I think we can come up with a way to not have to skip this test and just supply the correct arch to it without relying on the host

File: pkg/virt-launcher/virtwrap/manager_test.go (hunk 1 above)
Comment (vamsikrishna-siddu) [reply pkg/virt-launcher/virtwrap/manager_test.go]: Agreed. Updated the test to assert the expected rejection error on unsupported architectures instead of skipping.

File: pkg/virt-launcher/virtwrap/manager_test.go (hunk 2 above)
Comment (vamsikrishna-siddu) [reply pkg/virt-launcher/virtwrap/manager_test.go]: Decorators like WgS390x exist for E2E tests, but not for unit tests. The updated code addresses this by asserting the expected behaviour per architecture instead.

File: pkg/virt-launcher/virtwrap/manager_test.go [hunk 3]
```diff
@@ -3277,9 +3277,13 @@ var _ = Describe("Manager", func() {
 			libvirtManager.iommuFD = int(iommuFDFile.Fd())
 			converterContext, err := libvirtManager.generateConverterContext(vmi, true, options, false)
 
-			Expect(err).ToNot(HaveOccurred())
-			Expect(converterContext.GraceIOVirtualizationEnabled).To(BeTrue())
-			Expect(converterContext.GraceHostDeviceAliases).To(Equal([]string{"gpu-gpu0"}))
+			if arch.NewConverter(runtime.GOARCH).SupportPCIePlacement() {
```
Comment (lyarwood) [inline pkg/virt-launcher/virtwrap/manager_test.go:3280]: I'd rather see this broken out into two tables with entries per arch.

Comment (akalenyu) [inline pkg/virt-launcher/virtwrap/manager_test.go:3280]: I think the branching in the test makes it harder to reason about (sort of a positive/negative mixture);
To me, it would read more naturally if this test case would only **supply** the context it's interested in testing (`SupportPCIePlacement`) and a negative (separate) case (`!SupportPCIePlacement`) would exercise the error path.

**Comment by fossedihelm** [comment]
Let's merge this in the meantime. This is blocking everything from being merged
/lgtm
/unhold
/kind bug
